Sources/DeepLearning/Layer.swift

```diff
@@ -1295,3 +1295,125 @@ public struct Reshape<Scalar: TensorFlowFloatingPoint>: Layer {
+    /// - Parameters:
+    ///   - input: The input to the layer.
+    ///   - context: The contextual information for the layer application, e.g. the current learning
+    ///     phase.
+    /// - Returns: The hidden state.
+    @differentiable
+    public func applied(to input: Input, in _: Context) -> Tensor<Scalar> {
+        let concatenatedInput =  input.stepInput.concatenated(with: input.state)
+        return matmul(concatenatedInput, weight) + bias
+    }
+    
+    public func zeroState() -> Tensor<Scalar> {
+        return Tensor(zeros: stateShape)
+    }
+}
+
+/// An LSTM Cell.
+public struct LSTMCell<Scalar: TensorFlowFloatingPoint>: Layer {
+    public var inputWeight: Tensor<Scalar>
+    public var updateWeight: Tensor<Scalar>
+    public var forgetWeight: Tensor<Scalar>
+    public var forgetBias: Tensor<Scalar>
+    public var outputWeight: Tensor<Scalar>
+    
+    @noDerivative var stateShape: TensorShape
+
+    /// Creates a `LSTMCell` with the specified input size and hidden state size.
+    ///
+    /// - Parameters:
+    ///   - inputSize: The number of features in 2-D input tensors.
+    ///   - hiddenSize: The number of features in 2-D hidden states.
+    public init(inputSize: Int, hiddenSize: Int) {
+        let concatenatedInputSize = Int32(inputSize + hiddenSize)
+        let gateWeightShape = TensorShape([concatenatedInputSize, Int32(hiddenSize)])
+        self.inputWeight = Tensor(glorotUniform: gateWeightShape)
+        self.updateWeight = Tensor(glorotUniform: gateWeightShape)
+        self.forgetWeight = Tensor(glorotUniform: gateWeightShape)
+        self.forgetBias = Tensor(zeros: [Int32(hiddenSize)])
+        self.outputWeight = Tensor(glorotUniform: gateWeightShape)
+        self.stateShape = TensorShape([1, concatenatedInputSize])
+    }
+
+    public struct State: Differentiable {
+        public var cell: Tensor<Scalar>
+        public var hidden: Tensor<Scalar>
+
+        @differentiable
+        public init(cell: Tensor<Scalar>, hidden: Tensor<Scalar>) {
+            self.cell = cell
+            self.hidden = hidden
+        }
+    }
+
+    public struct Input: Differentiable {
+        public var stepInput: Tensor<Scalar>
+        public var state: State
+        
+        @differentiable
+        public init(stepInput: Tensor<Scalar>, state: State) {
+            self.stepInput = stepInput
+            self.state = state
+        }
+    }
+
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameters:
+    ///   - input: The input to the layer.
+    ///   - context: The contextual information for the layer application, e.g. the current learning
+    ///     phase.
+    /// - Returns: The hidden state.
+    @differentiable
+    public func applied(to input: Input, in _: Context) -> State {
+        let gateInput = input.stepInput.concatenated(with: input.state.hidden, alongAxis: 1)
+
+        let inputGate = sigmoid(matmul(gateInput, inputWeight))
```

Comment:
As a further optimization you could fuse all gates' weights into a single Tensor and do a single matmul:

```swift
let (input, update, forget, output) = split(matmul(gateInput, weight)) + bias, 4)  # Wishful thinking.
...
```

Reply:
Agreed, I just thought it was more understandable/readable this way. What do you think @rxwei?

Reply:
Definitely so, but we do not have a Swifty implementation of the `Split` op yet.

Reply:
I'd rather leave it exploded out and rely on XLA/other optimizations to make it more efficient later.